### PR TITLE
[Java] [XSUAA] Introduction & Usage

### PR DESCRIPTION
**Pull request by newtork**
In our team there is uncertainty on **how to obtain/use access tokens**. In my opinion this knowledge "is important" because otherwise you will not be able to explore and test your assumptions on services responses on SCP CF. The main driver for this document was the recent introduction of _new destination authorization types_, that required some manual investigation first.

My goal is to share knowledge:
- [ ] What are access tokens? What kind of token is important for what use case?
- [ ] Where do access tokens come from? How does the token reach my application request header?
- [x] How do obtain my own access token?
- [ ] How can I test services on SCP CF with that token?

## Context

I think it would be good to have an introductory document on how to deal with access tokens and client credentials on SCP CF.

This contains a small collection of useful requests/responses with the XSUAA service endpoint.

## Definition of Done

- [ ] If applicable: Check if `node run doc` still works.

### CODE REVIEW

File: docs/java/guides/cf-xsuaa.md [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+- cloud native
+- cloud sdk
+- sap cloud sdk
+- xsuaa
+- destination
+- connectivity
+image:
+---
+
+## XSUAA Usage on Cloud Foundry
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:22]:
```suggestion
## XSUAA Usage on SCP Cloud Foundry
```

File: docs/java/guides/cf-xsuaa.md [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+When implementing and deploying an application it quickly becomes important to understand how authentication and authorizations works on SCP Cloud Foundry.
+In our tutorials and documentation pages we recommend using "approuter" as proxy service to handle authentication management to your implemented application.
+The following steps will show what happens behind the scenes. The requests can be manually reproduced by a REST client of your choice, e.g. Postman or Insomnia.
+
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:27]:
```suggestion
When developing and deploying an application it quickly becomes important to understand how authentication and authorizations work on SCP Cloud Foundry.
In our tutorials and documentation, we recommend using "approuter" as a proxy service to handle authentication management to your implemented application.
The following steps will show what happens behind the scenes. The requests can be manually reproduced by a REST client of your choice, e.g. Postman or Insomnia.

```

File: docs/java/guides/cf-xsuaa.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+The [official UAA documentation](https://docs.cloudfoundry.org/api/uaa/) is highly recommended if there are questions.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:28]:
```suggestion
We highly recommend checking the [official UAA documentation](https://docs.cloudfoundry.org/api/uaa/) for any questions and comprehensive documentation on the topic.
```

File: docs/java/guides/cf-xsuaa.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+
+**Please note:** The following paragraphs only describe a subset of features of the XSUAA Service on Cloud Foundry.
+The Cloud SDK is not participating in the active development of this software and cannot give in-depth support.
+Some information may already be outdated.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:32]:
```suggestion
:::note XSUAA service is developed independently of SAP Cloud SDK
The following documentation only touches a subset of features of the XSUAA Service on Cloud Foundry.
The Cloud SDK and XSUAA are developed independently. We do not provide in-depth support on XSUAA topics beyond Cloud SDK use cases. Mind, if some information seems outdated - get in touch with us and refer to [official XSUAA docs]((https://docs.cloudfoundry.org/api/uaa/)).
:::
```

File: docs/java/guides/cf-xsuaa.md [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+
+These are use cases described below:
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:35]:
```suggestion
The Cloud SDK handles the following use-cases:
```

File: docs/java/guides/cf-xsuaa.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+- User Login: `Authorization Code Grant`
+- SCP Service Usage on behalf of User: `JWT Bearer Token Grant`
+- SCP Service Usage on behalf of service: `Client Credentials Grant`
+- Resolve User Access Token: `Refresh Token Grant`
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:39]:
```suggestion
- User Login: `Authorization Code Grant`
- SCP Service Usage on behalf of a User: `JWT Bearer Token Grant`
- SCP Service Usage on behalf of a service: `Client Credentials Grant`
- Resolve User Access Token: `Refresh Token Grant`
```

File: docs/java/guides/cf-xsuaa.md [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+### Read the Application Properties  
+
+In order to create a request, we need to parse the XSUAA connection data.
+
+1. First, take note of your "_application route_", that's the URL for which an authorization request will be constructed.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:45]:
```suggestion
1. Take note of your "_application route_". That's the URL for which an authorization request will be built.
```

File: docs/java/guides/cf-xsuaa.md [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+1. Open the **system-provided** environment variables of your application on Cloud Foundry.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:47]:
```suggestion
2. Open the **system-provided** environment variables of your application on Cloud Foundry.
```

File: docs/java/guides/cf-xsuaa.md [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+1. In the JSON value, locate the object `VCAP_SERVICES.xsuaa[0].credentials`.
+   
+   *Note:* Depending on the scenario, the "xsuaa" array may have more than one entry. Your application can be bound to multiple instances, e.g. through different service plans.
+
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:52]:
```suggestion
3. Extract values "_url_", "_clientid_", "_clientsecret_" from the JSON value, locate in the object `VCAP_SERVICES.xsuaa[0].credentials`.
   
:::tip
Depending on your setup, the `xsuaa` array may have more than one entry. Because your application can be bound to multiple instances, e.g. through different service plans.
:::
```

File: docs/java/guides/cf-xsuaa.md [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+1. Extract values "_url_", "_clientid_", "_clientsecret_".
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:53]:
```suggestion
```

File: docs/java/guides/cf-xsuaa.md [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+
+## Authorization Code Grant
+
+Since we start without previous access token, we begin our journey with the browser flow of [Authorization Code Grant](https://docs.cloudfoundry.org/api/uaa/version/74.23.0/index.html#authorization-code-grant).
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:57]:
```suggestion
Since we start without an existing access token, our journey begins with the browser flow of [Authorization Code Grant](https://docs.cloudfoundry.org/api/uaa/version/74.23.0/index.html#authorization-code-grant).
```

File: docs/java/guides/cf-xsuaa.md [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+
+The following action is split into two steps:
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:59]:
```suggestion
This flow is split into two steps: 
```

File: docs/java/guides/cf-xsuaa.md [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+- Get authorization code on behalf of single-sign-on login form.
+- Get personal access token from authorization code.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:61]:
```suggestion
- Get authorization code on behalf of a single-sign-on login form.
- Get personal access token from authorization code.
```

File: docs/java/guides/cf-xsuaa.md [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+
+### Get Authorization Code
+
+You will likely need to run the following request in your browser and monitor the HTTP response.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:65]:
```suggestion
You will likely need to run the following HTTP request in your browser and check the HTTP response.
```

File: docs/java/guides/cf-xsuaa.md [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+
+1. Request:
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:67]:
```suggestion
1. HTTP Request example:
```

File: docs/java/guides/cf-xsuaa.md [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+    ```
+    GET https://[xsuaa.url]/oauth/authorize
+    
+    Query parameters:
+    
+    client_id=[xsuaa.clientid]
+    redirect_uri=[application.route]
+    response_type=code
+    ```
+    (optional values can be set for "scope" and "login_hint")
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:77]:
```suggestion
:::tip
Optional values can be set for "scope" and "login_hint"
:::
```

File: docs/java/guides/cf-xsuaa.md [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+1. Submit login form via browser.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:78]:
```suggestion
2. Submit login form via a browser or REST API debugging tools like POSTMAN or Insomnia.
```

File: docs/java/guides/cf-xsuaa.md [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+2. Read HTTP response:
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:79]:
```suggestion
3. Check the HTTP response and extract `[code]` from `Location` header.
```

File: docs/java/guides/cf-xsuaa.md [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+    Strict-Transport-Security: max-age=31536000
+    Set-Cookie: X-Uaa-Csrf=[...]; Path=/; Max-Age=86400; Expires=[...]; HttpOnly
+    Cache-Control: no-store
+    Content-Language: en
+    Location: [application.route]?code=[code]
+    X-XSS-Protection: 1; mode=block
+    X-Frame-Options: DENY
+    X-Content-Type-Options: nosniff
+    ```
+1. Extract `[code]` from `Location` header.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:91]:
```suggestion
```

File: docs/java/guides/cf-xsuaa.md [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+
+### Get OAuth2 Access Token
+
+With the authorization code we can now request a real access token from the OAuth2 service endpoint:
+
+1. Request:
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:97]:
```suggestion
1. Make the following `Request`:
```

File: docs/java/guides/cf-xsuaa.md [hunk 21]
```diff
@@ -0,0 +1,229 @@
...
+    Accept: application/json
+    Content-Type: application/x-www-form-urlencoded
+   
+    client_id=[xsuaa.clientid]
+    client_secret=[xsuaa.clientsecret]
+    redirect_uri=[application.route]
+    code=[code]
+    grant_type=authorization_code
+    ```
+1. Response:
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:111]:
```suggestion
2. Check the `Response`:
```

File: docs/java/guides/cf-xsuaa.md [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+      "access_token": [access_token],
+      "token_type": "bearer",
+      "id_token": [...],
+      "refresh_token": [refresh_token],
+      "expires_in": [...],
+      "scope": [...],
+      "jti": [...]
+    }
+    ```
+   Congratulation, you have a valid `access_token`.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:123]:
```suggestion
   Congratulation, now you've fetched a valid `access_token`.
```

File: docs/java/guides/cf-xsuaa.md [hunk 23]
```diff
@@ -0,0 +1,229 @@
...
+   It can be further evaluated and forwarded.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:124]: This should be made more obvious. I don't fully get it.

File: docs/java/guides/cf-xsuaa.md [hunk 24]
```diff
@@ -0,0 +1,229 @@
...
+   
+   Some applications like _approuter_ will save the `refresh_token` to the user session.
+   This leverages the automatic retrieval of new access tokens, if the previous ones have expired during the active session.
+
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:128]:
```suggestion
:::tip
   Some applications like _approuter_ will save the `refresh_token` to the user session for you.
   This enables automatic retrieval of new access tokens after the existing one has expired during the active session.
:::
```

File: docs/java/guides/cf-xsuaa.md [hunk 25]
```diff
@@ -0,0 +1,229 @@
...
+## JWT Bearer Token Grant
+
+Several services on SCP Cloud Foundry require a dedicated OAuth2 access token, e.g. _Connectivity Service_ and _Destination Service_.
+
+1. Open the **system-provided** environment variables of your application.
+
+1. In the JSON value, locate the object `VCAP_SERVICES.destination[0].credentials`.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:135]:
```suggestion
2. In the JSON value, locate the object `VCAP_SERVICES.destination[0].credentials`. Make note of `clientid`, `clientsecret`, `uri`
```

File: docs/java/guides/cf-xsuaa.md [hunk 26]
```diff
@@ -0,0 +1,229 @@
...
+
+1. Read "clientid", "clientsecret", "uri".
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:137]:
```suggestion
```

File: docs/java/guides/cf-xsuaa.md [hunk 27]
```diff
@@ -0,0 +1,229 @@
...
+   
+1. Request:
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:139]:
```suggestion
3. Make the following `Request`:
```

File: docs/java/guides/cf-xsuaa.md [hunk 28]
```diff
@@ -0,0 +1,229 @@
...
+    Accept: application/json
+    Content-Type: application/x-www-form-urlencoded
+   
+    client_id=[destination.clientid]
+    client_secret=[destination.clientsecret]
+    assertion=[access_token]
+    grant_type=urn:ietf:params:oauth:grant-type:jwt-bearer
+    response_type=token
+    ```
+1. Response:
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:153]:
```suggestion
4. Check the `Response`:
```

File: docs/java/guides/cf-xsuaa.md [hunk 29]
```diff
@@ -0,0 +1,229 @@
...
+    {
+      "access_token": [destination_access_token],
+      "token_type": "bearer",
+      "expires_in": [...],
+      "scope": [...],
+      "jti": [...]
+    }
+   ```
+   Congratulation, you have a valid `destination_access_token`.
+   It can be used to query the `destination.uri` linked destination service on behalf of the current user.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:164]:
```suggestion
   Congratulation, you have a valid `destination_access_token`.
   It can be used to query the `destination.uri` linked to destination service on behalf of the current user.
```

File: docs/java/guides/cf-xsuaa.md [hunk 30]
```diff
@@ -0,0 +1,229 @@
...
+
+## Client Credentials Grant
+
+Some applications require access to a platform service without an active user sessions, with _technical user_ credentials.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:168]:
```suggestion
Some applications require access to a platform service without active user sessions. This requires credentials known as _technical user_ credentials.
```

File: docs/java/guides/cf-xsuaa.md [hunk 31]
```diff
@@ -0,0 +1,229 @@
...
+For example, reading a list of destinations does not require a user access token.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:169]:
```suggestion
For example, reading a list of `destinations` does not require a `user access token`.
```

File: docs/java/guides/cf-xsuaa.md [hunk 32]
```diff
@@ -0,0 +1,229 @@
...
+Instead we can request an access token on behalf of the service binding itself.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:170]:
```suggestion
Instead, we can request an `access token` on behalf of the service binding itself.
```

File: docs/java/guides/cf-xsuaa.md [hunk 33]
```diff
@@ -0,0 +1,229 @@
...
+Here we use the [Client Credentials Grant](https://docs.cloudfoundry.org/api/uaa/version/74.23.0/index.html#without-authorization).
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:171]:
```suggestion
In the example below, we'll use the [Client Credentials Grant] (https://docs.cloudfoundry.org/api/uaa/version/74.23.0/index.html#without-authorization) to get `access token`.
```

File: docs/java/guides/cf-xsuaa.md [hunk 34]
```diff
@@ -0,0 +1,229 @@
...
+
+1. Request:
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:173]:
```suggestion
1. Make a `Request`:
```

File: docs/java/guides/cf-xsuaa.md [hunk 35]
```diff
@@ -0,0 +1,229 @@
...
+    
+    Headers
+    Accept: application/json
+    Content-Type: application/x-www-form-urlencoded
+   
+    client_id=[destination.clientid]
+    client_secret=[destination.clientsecret]
+    grant_type=client_credentials
+    ```
+1. Response:
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:185]:
```suggestion
2. Check the `Response`:
```

File: docs/java/guides/cf-xsuaa.md [hunk 36]
```diff
@@ -0,0 +1,229 @@
...
+    {
+      "access_token": [destination_access_token],
+      "token_type": "bearer",
+      "expires_in": [...],
+      "scope": [...],
+      "jti": [...]
+    }
+    ```
+   Congratulation, you have a valid `destination_access_token`.
+   It can be used to query the `destination.uri` linked destination service on behalf of the service binding itself.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:196]:
```suggestion
   It can be used to query the `destination.uri` linked destination service on behalf of the service binding.
```

File: docs/java/guides/cf-xsuaa.md [hunk 37]
```diff
@@ -0,0 +1,229 @@
...
+
+   
+   
+## Refresh Token Grant
+
+If the current access token is expired, a new one can be requested with the [Refresh Token](https://docs.cloudfoundry.org/api/uaa/version/74.23.0/index.html#refresh-token).
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:202]:
```suggestion
If the current access token is expired, a new one can be requested with the [Refresh Token flow](https://docs.cloudfoundry.org/api/uaa/version/74.23.0/index.html#refresh-token).
```

File: docs/java/guides/cf-xsuaa.md [hunk 38]
```diff
@@ -0,0 +1,229 @@
...
+   
+1. Request:
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:204]:
```suggestion
1. Make a `Request`:
```

File: docs/java/guides/cf-xsuaa.md [hunk 39]
```diff
@@ -0,0 +1,229 @@
...
+    Headers
+    Accept: application/json
+    Content-Type: application/x-www-form-urlencoded
+   
+    client_id=[xsuaa.clientid]
+    client_secret=[xsuaa.clientsecret]
+    refresh_token=[refresh_token]
+    grant_type=refresh_token
+    ```
+1. Response:
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:217]:
```suggestion
2.Check the `Response`:
```

File: docs/java/guides/cf-xsuaa.md [hunk 40]
```diff
@@ -0,0 +1,229 @@
...
+      "access_token": [access_token],
+      "token_type": "bearer",
+      "id_token": [...],
+      "refresh_token": [refresh_token],
+      "expires_in": [...],
+      "scope": [...],
+      "jti": [...]
+    }
+    ```
+   Congratulation, you have a refreshed `access_token`.
```
Comment (artemkovalyov) [inline docs/java/guides/cf-xsuaa.md:229]:
```suggestion
   Congratulation, you now have a refreshed `access_token`.
```

**Comment by FrankEssenberger** [comment]
This is very valuable and condense information. I think this is not language specific so the [Java] should be removed and we should also add it to the general documentaion. Since @mr-flannery is our XSUAA expert perhaps he can have a look. 

I had. a quick look and for me it would be good to have two sentences on each type of grant pointing out the usecase like: user grant contains the user from the jwt token encoded. Client grant use unspecific used for app to app communication. There is already something there but a bit more would be nice. 

Also I missed the point about tenat specific token. I think the tenant sub domain is then used in the xsuaa url instead of the inital one. Perhaps a paragraph to illustrate this would make the document even more complete.

**Comment by mr-flannery** [comment]
The things described in the document look sound to me. So I'll leave you with some general thoughts:

What's the intended scope of the document? Currently the document is motivated with (paraphrased) "it becomes important to understand how authentication and authorizations work". I agree with the statement, but I'm not sure whether this document would fully answer these questions, since there are so many more things to talk about. For example, this document is rather focused on the non-SAP parts (OAuth flows, login via SAML token exchange (though I think SAML is not explicitly mentioned IIRC)), but doesn't mention multi-tenancy, for example. So depending on the reader's context, this might be confusing (?).

Of course, the content that is there is a great start already, so maybe it's a good idea to explicitly state what is covered in this document and what is not covered (but might be in the future?).

Also, there is existing documentation about both the XSUAA service as well as the approuter, as far as I'm aware. Have you thought about what the benefit of this documentation is compared to the existing documentation?

Hope that helps! (Otherwise ignore my comments 😄 )

File: docs/java/guides/cf-xsuaa.md (hunk 5 above)
Comment (newtork) [reply docs/java/guides/cf-xsuaa.md:36]: The Cloud SDK does not handle all those cases:

- [ ] User Login: `Authorization Code Grant` (approuter does it)
- [x] SCP Service Usage on behalf of a User: `JWT Bearer Token Grant`
- [x] SCP Service Usage on behalf of a service: `Client Credentials Grant`
- [ ] Resolve User Access Token: `Refresh Token Grant` (approuter does it)

**Comment by artemkovalyov** [comment]
> The things described in the document look sound to me. So I'll leave you with some general thoughts:
> 
> What's the intended scope of the document? Currently the document is motivated with (paraphrased) "it becomes important to understand how authentication and authorizations work". I agree with the statement, but I'm not sure whether this document would fully answer these questions, since there are so many more things to talk about. For example, this document is rather focused on the non-SAP parts (OAuth flows, login via SAML token exchange (though I think SAML is not explicitly mentioned IIRC)), but doesn't mention multi-tenancy, for example. So depending on the reader's context, this might be confusing (?).
> 
> Of course, the content that is there is a great start already, so maybe it's a good idea to explicitly state what is covered in this document and what is not covered (but might be in the future?).
> 
> Also, there is existing documentation about both the XSUAA service as well as the approuter, as far as I'm aware. Have you thought about what the benefit of this documentation is compared to the existing documentation?
> 
> Hope that helps! (Otherwise ignore my comments  )

File: docs/java/guides/cf-xsuaa.md (hunk 17 above)
Comment (newtork) [reply docs/java/guides/cf-xsuaa.md:81]: I doubt this works with postman / insomnia, because this step requires a real user session and likely cookie support (for single-sign-on). IMHO this is harder to describe than just telling the reader: use the browser.

**Comment by newtork** [comment]
I should elaborate on my motivation:

In our team there is uncertainty on **how to obtain/use access tokens**. In my opinion this knowledge "is important" because otherwise you will not be able to explore and test your assumptions on services responses on SCP CF. The main driver for this document was the recent introduction of _new destination authorization types_, that required some manual investigation first.

My goal is to share knowledge:
- [ ] What are access tokens? What kind of token is important for what use case?
- [ ] Where do access tokens come from? How does the token reach my application request header?
- [x] How do obtain my own access token?
- [ ] How can I test services on SCP CF with that token?


@FrankEssenberger wrote
>I had. a quick look and for me it would be good to have two sentences on each type of grant pointing out the usecase like: user grant contains the user from the jwt token encoded. Client grant use unspecific used for app to app communication. There is already something there but a bit more would be nice.

You are right. It could use a little more detail on what **is** an _access token_. Although this shouldn't be the main focus on the document. JWT

> Also I missed the point about tenat specific token. I think the tenant sub domain is then used in the xsuaa url instead of the inital one. Perhaps a paragraph to illustrate this would make the document even more complete.

Yeah, that would be cool. But I have a feeling, because of the explanation/complexity this would justify a separate PR for expanding the document.

@mr-flannery wrote:
> What's the intended scope of the document?

Good point, I need to limit the described scope of the document and provide a proper introduction. At least on the chapter that I'm trying to provide.


> For example, this document is rather focused on the non-SAP parts (OAuth flows, login via SAML token exchange (though I think SAML is not explicitly mentioned IIRC)), but doesn't mention multi-tenancy, for example. So depending on the reader's context, this might be confusing (?).

AFAIK the [SAML2 bearer grant](https://docs.cloudfoundry.org/api/uaa/version/74.4.0/index.html#saml2-bearer-grant) is deprecated and [JWT bearer token grant](https://docs.cloudfoundry.org/api/uaa/version/74.23.0/index.html#jwt-bearer-token-grant) should be preferred. That's what the XSUAA client lib colleagues said.

I would like to cover multi-tenancy in a separate PR (for this document).

> Also, there is existing documentation about both the XSUAA service as well as the approuter, as far as I'm aware. Have you thought about what the benefit of this documentation is compared to the existing documentation?

I wanted a boiled down document on what exact HTTP request I have to run on SCP CF in order to gain an access token. To my feeling, the existing remote documentation is too crowded / not detailed enough.

**Comment by newtork** [comment]
@mr-flannery wrote:
> What's the intended scope of the document?

Good point, I need to limit the described scope of the document and provide a proper introduction. At least on the chapter that I'm trying to provide.


> For example, this document is rather focused on the non-SAP parts (OAuth flows, login via SAML token exchange (though I think SAML is not explicitly mentioned IIRC)), but doesn't mention multi-tenancy, for example. So depending on the reader's context, this might be confusing (?).

AFAIK the [SAML2 bearer grant](https://docs.cloudfoundry.org/api/uaa/version/74.4.0/index.html#saml2-bearer-grant) is deprecated and [JWT bearer token grant](https://docs.cloudfoundry.org/api/uaa/version/74.23.0/index.html#jwt-bearer-token-grant) should be preferred. That's what the XSUAA client lib colleagues said.

I would like to cover multi-tenancy in a separate PR (for this document).

> Also, there is existing documentation about both the XSUAA service as well as the approuter, as far as I'm aware. Have you thought about what the benefit of this documentation is compared to the existing documentation?

I wanted a boiled down document on what exact HTTP request I have to run on SCP CF in order to gain an access token. To my feeling, the existing remote documentation is too crowded / not detailed enough.

**Comment by artemkovalyov** [comment]
@mr-flannery, I agree on reasoning from @newtork  about having condensed info about the topic raising in every cloud App from the SDK perspective. It also makes things available fro SDK devs from one place and provides reverence to relevant sources we are guided by. This is a huge value and justifies the effort. I also agree that if you're starting from scratch in SAP ecosystems you'll be defeated by the number of docs and sources, so having guidance is of crucial importance for SDK adopters, especially external audience.

Thanks for keeping us in check with added value perspective anyway. I'd say code snippets or some logs from real-life APPs might be a valuable addition in the future.

File: docs/java/guides/cf-xsuaa.md [hunk 41]
```diff
@@ -0,0 +1,231 @@
+---
+id: cf-xsuaa
+title: SAP Cloud Foundry
+hide_title: false
+hide_table_of_contents: false
+sidebar_label: Cloud Foundry XSUAA Usage
+custom_edit_url: https://github.com
+description: Configure you Cloud Foundry CLI and bind it to SAP Cloud Platform
```
Comment (MatKuhr) [inline docs/java/guides/cf-xsuaa.md:8]:
```suggestion
description: Understand how the XSUAA service on Cloud Foundry works
```

File: docs/java/guides/cf-xsuaa.md (hunk 8 above)
Comment (MatKuhr) [reply docs/java/guides/cf-xsuaa.md:48]: Actually having the `1.` everywhere makes sense. That way you don't have to edit anything when adding or deleting steps.

File: docs/java/guides/cf-xsuaa.md [hunk 42]
```diff
@@ -0,0 +1,231 @@
...
+- cloud native
+- cloud sdk
+- sap cloud sdk
+- xsuaa
+- destination
+- connectivity
+image:
+---
+
+## XSUAA Usage on SCP Cloud Foundry
```
Comment (MatKuhr) [inline docs/java/guides/cf-xsuaa.md:22]: I think we should make it more prominent that this guide is completely unrelated to the Cloud SDK. We don't want users to think that they have to do all of these things when using the SDK. 

So maybe use a headline like `Background knowledge on Authorization & Authentication`. And a prominent note that the Cloud SDK handles all these things and this is  mostly FYI.

**Comment by MatKuhr** [comment]
> We could consider adding a note specifically stating that these are just our notes and when in doubt always the official documentation should be relied upon. Otherwise consumers could rightfully blame us if something doesn't work as described.

Nevermind I see that note already exists, I just overlooked it.
